casanovo/denovo/model.py

```diff
@@ -444,6 +461,37 @@ def _terminate_finished_beams(
                             abs(d) < self.precursor_mass_tol
                             for d in delta_mass_ppm
                         )
+
+                        # Don't terminate if m/z difference is by NH3 loss
```

Comment:
**issue:** This solution seems sub-optimal because it considers NH3 loss as a special token, whereas there are different N-terminal modifications as well. Why did NH3 loss get special treatment? I adds a decent amount of extra complexity and I don't understand why this is needed (not documented).

Reply:
NH3 loss is the only token with negative mass. In the previous implementation, early stopping kicked in the even if the beam was expecting a NH3 loss as the last token because m/z tolerance would be exceeded. I couldn't think of a more elegant way of covering this edge case and avoiding m/z threshold based stopping if the m/z margin is equivalent to NH3 loss

Reply:
Ok, I didn't think/realize that NH3 loss is the only negative mass. Then it is indeed a bit of a special case, but we can try to generalize the code a little bit after the remaining comments have been addressed.